### ISSUE TITLE
Implement an updated logging (Spec layout and discussion)

### ISSUE THREAD

**Issue by IBurn36360** [OpenAngelArena/oaa#1443]
This is a list of general thoughts and a base spec layout for a logging API both internal to the game/server client side and the server side where the API sends the event payloads to.

# General Thoughts

## Game Side
- Internal API should be flexible and allow payloads to be generated for the following cases:
    - Informational payloads, such as general number or status based events not necessarily associated with any other event or error
    - Status payloads, such as a start of game event, duel start event etc.
    - Sync payloads, such as sync errors or anything specific with the network handling and updating before the main game loop takes control of the data (These are very often, so we would quantify what counts as a sync error and likely throttle them)
    - Exception payloads, such as caught exceptions that may or may not contribute to game breaks or errors.  In this case, exception should be reserved for any error that is caught and handled.
    - Error payloads, such as uncaught exceptions that may or may not contribute to game breaks or errors.  The alternate case of the above
    - Fatal payloads, sent on game hard crash, if possible.  In the event a game sends no events for a reasonable amount of time to the API, the server should consider the game as fatal for one reason or another.  A host disconnect or lobby disconnect should be flagged as such if possible.
- The API CAN NOT collide with game performance, meaning all API interaction and payload sending SHOULD be as light as possible and SHOULD attempt to allow the game to proceed before its own processing.  Any API that affects the gameplay in a detrimental manner will drive users away and should be avoided at all costs.
- The API should be available in all scopes as a global or superglobal and should present a simple and unified interface for developers to use.
- The API should collect backtraces for any and all errors, when available, on all event of severity Exception and higher (Sync might be a candidate for this as well).  Events may request a trace on behalf of the payload for events who do not already get backtraces.
- The API format MUST allow for dynamic allocation of data, that the view on the server side can decide on how to display later.  This is to prevent an event from being sendable due to design limitations and allow for deep level data collection where it makes sense.

## Server Side (Non-DOTA)
- Ingest point for log payloads should be resolvable for all people (Ideally should be a domain name pointed to a server or set of servers)
- Endpoint should validate requests are either a chain creation or are part of an existing chain (Prevent malicious or orphaned events from going into the system)
- Endpoint should be as fast as possible to release the connection so a new one can be made by the HTTP daemon to prevent load issues when events like a tournament are going on.
- Views should understand how to bucket and display the log events in a dashboard for easy at-a-glance vieweing
- Views may or may not require authentication, so allow all views to have permission controls
- Views should NOT affect the API functionality, so long queries are to be avoided if at all possible
- Ingest should work somewhat like how the Loggly software works in design and display.

# Spec

## Game Client
```php
// The following is a proposal to the internal API structure of the logging layer and is subject to change

/**
 * Dispatches a log event to the server
 *
 * @param int eventSeverity - The integer event severity.  Constants in the OAALogging struct should be used to define event severity for ease-of-use and ease-of-upgrade
 * @param string eventDescription - The string short description of the event.  This should concisely describe the event so a developer later on can determine if the event should be looked at at a glance.  Should be capped at 200 characters.
 * @param mixed|object= - The optional event payload.  May be of any type of collection, such an an object, table or array and is to be stored in a serialized manner for later use on the server side
 *
 * @access public
 *
 * @return void
 */
{Global space}.OAALogging.log(int eventSeverity, string eventDescription, eventPayload = []) { /* ... */ }
```

## Server API
The following is a proposal to the server API structure of the logging layer and is subject to change

### Endpoint (Ingest)
- Requirements
    - Request payload MUST include the following post-body fields
        - eventSeverity
        - eventMessage
        - eventPayload
    - Request MUST be sent in a decipherable POST body format, such as HTTP form-encoded or JSON.
    - Request body MUST be free of any other fields.  Any other fields detected should result in the API rejecting the request (This can be used on test clients to nullify the API so events are not sent to production if that is a desired effect)
    - Headers MUST contain the following fields
        - X-OAA-Check header containing the SHA256 or SHA512 hash of the payload data (Before encoding.  Specific details on how this is handled are not determined in initial spec, but are suggested to be a simple serialization of all fields and values into a single string for hash)
            - Failing to have this header, or to have this header value NOT match the calculated hash of the body is to result in the request being rejected
        - X-OAA-ST header containing the UNIX timestamp start time.  This SHOULD match up with the sent start time of the payload.  If the payload time is out of range, say 30 or more seconds out of date, or the timestamp is out of tolerance to the actual send time on the payload (1 second variance)is to result in the request being rejected
        - X-OAA-GUID header containing the game UID
            - This should always be provided, if possible.  Not knowing if this is possible at the moment bars this from being a spec requirement
        - X-OAA-Version header containing the OAA version
            - This should always be provided, but is not being required in initial spec
        - X-OAA-ClientVersion header containing the client version
            - This should always be provided, if possible.  Not knowing if this is possible at the moment bars this from being a spec requirement
    - Successful requests that are ingested MUST send back a proper HTTP 200 (OK) response with an optional response body message, if there is any desire for the client side to know what response message was sent.
    - Failed API requests MUST send back a proper 501 (Server Error) response with an optional response body message.
    - Rejected API requests MUST send back a proper 400 (Bad Request) response with an optional response body message detailing the rejection reason.

### Considerations
- Should there be a separated section for developers to have their DOTA Tools games logged?
- Should there be considerations to allow anonymous access to logs?
- Should we need to batch up some events and sent them to the API in a larger payload?
- Should there be a maximal limitation on how large payloads can be?
- Should there be a minimum requirement for a payload to avoid sending frivolous information to the server?

**Comment by chrisinajar** [comment]
Still haven't read all of this, it'd be nice to have something like loggly on crack, but it'd be even nicer to have something like Crashlitics.

Error reporting is super important, and we actually have a bunch of custom code for reading in these errors already. We can pretty easily hook that logic into a system sending and batching together similar errors, then we can see what % of sessions have them and perhaps check out the event stream leading up to the crash.

also loggly

**Comment by IBurn36360** [comment]
I'm splitting this issue into 2 distinct parts, application (DOTA client and Valve dedicated server) and reporting (Endpoint and views system for displaying the collected data).

The application side describes how, internally, data will be sent to the reporting server (In an abstract enough manner that allows for flexibility and easy pivoting).  

The views side is going to take what I have been finding out of Loggly and tailoring what it does specifically for the use of the data collected from OAA games.  The specifics of that side aren't detailed as they rely entirely on knowing what data is going to be captured.

**Comment by IBurn36360** [comment]
Adding an update to this so it doesn't look abandoned.

- LUA side is loading (It has been so long since I used LUA that it took a while for me to even get used to the syntax again, but all is good now).  Working on finding documentation on the HTTPRequestClientVM's HTTPRequestHandle object to determine what the reporting server API requirements will need to be (For instance, does the API need to drop all header requirements and attempt to decypher the request coming in on its own.  This will be determined by just how usable the HTTPClient is)
- Server side doesn't have work done (Not that it will be that hard for me to get a script up to accept the API requests and insert them into the DB securely), but plan is as follows:
    - Once the game-side work is done and the tests using my reflection API setup on my site are completed, set up a PR with that addition and discuss what logging items in the current game should be migrated to the new internal to external logging system (First in a parallel manner, where we would keep both the console logging and implement the to-external logging.  Once we confirm the game is both running 100% on the external reporting and that it has no possible normal exceptions, we would then remove the virtual console logging until the logging fails (in which the in-built systems that would still exist would kick in and send the exception to the console))
    - Once we are running on reporting, then the usability of the views system, which the discussion on how that should be laid out should start now, then specified views for viewing any information will be built as needs arise (We should strive to almost log too much information, as having as much information as possible will allow for better usability out of the reporting system)
        - Questions for the views side on the reporting servers
            - Is there any need to have these logs locked behind authentication?
                - If yes, is there any need to expand this to group-based permissions, or just delegated users?
            - Does this need to integrate with any website framework? (Such as WordPress, Drupal, PHPbb, etc.)
                - This may force decisions for the above, so it may be pertinent to answer this before the above
            - Does hosting for this already exist in a form where we can setup and manage a database? (If not, I have servers, so that is no problem)
             - Does this live under the openangelarena.com domain, or does it need to go to a separated domain?
